web/src/shared/components/Graphs/HorizontalStackedBarChart.tsx

```diff
@@ -0,0 +1,350 @@
+import React from 'react'
+import * as d3 from 'd3'
+import LoadingDucks from '../LoadingDucks/LoadingDucks'
+import { BillingCostBudgetRecord } from '../../../sm-api'
+import { grey } from '@mui/material/colors'
+
+interface HorizontalStackedBarChartProps {
+    data: BillingCostBudgetRecord[]
+    title: string
+    series: string[]
+    labels: string[]
+    total_series: string
+    threshold_values: number[]
+    threshold_series: string
+    sorted_by: string
+    colors?: (t: number) => string | undefined
+    isLoading: boolean
+    showLegend: boolean
+}
+
+const HorizontalStackedBarChart: React.FC<HorizontalStackedBarChartProps> = ({
+    data,
+    title,
+    series,
+    labels,
+    total_series,
+    threshold_values,
+    threshold_series,
+    sorted_by,
+    colors,
+    isLoading,
+    showLegend,
+}) => {
+    if (!isLoading && (!data || data.length === 0)) {
+        return <div>No data available</div>
+    }
+
+    const colorFunc: (t: number) => string | undefined = colors ?? d3.interpolateRainbow
+
+    // set the dimensions and margins of the graph
+    const margin = { top: 80, right: 20, bottom: 50, left: 250 }
+    const width = 650 - margin.left - margin.right
+    const outsideHeight = 2850
+    const height = outsideHeight - margin.top - margin.bottom
+
+    const containerDivRef = React.useRef<HTMLDivElement>()
+
+    const [clientWidth, setClientWidth] = React.useState(650)
+
+    React.useEffect(() => {
+        function updateWindowWidth() {
+            setClientWidth(containerDivRef.current?.clientWidth ?? 650)
+        }
+        if (containerDivRef.current) {
+            updateWindowWidth()
+        }
+        window.addEventListener('resize', updateWindowWidth)
+
+        return () => {
+            window.removeEventListener('resize', updateWindowWidth)
+        }
+    }, [])
+
+    const contDiv = containerDivRef.current
+    if (contDiv) {
```

Comment:
Was there a more react way to do this?

Reply:
Yeah, would need to revisit this one as well. Leave it for the time being.